web/src/copy_messages.ts

```diff
@@ -129,42 +129,84 @@ function get_nearest_html_element(node: Node | null): Element | null {
+    annotation element. This ensures that pasting functions as expected.
+
+    The primary objective of this expansion is to include the `<annotation>` tag within
+    the selection range so that it can be properly utilized by `paste_handler_converter`
+    as `paste_html`.
 */
 function improve_katex_selection_range(selection: Selection): void {
     const anchor_element = get_nearest_html_element(selection.anchorNode);
     const focus_element = get_nearest_html_element(selection.focusNode);
-    // If the anchor and focus end up in different katex spans, this selection
-    // isn't meant to be an inline expression, so we perform an early return.
+
     if (
         focus_element &&
         anchor_element &&
-        focus_element?.closest(".katex") !== anchor_element?.closest(".katex")
+        focus_element?.closest(".katex") !== anchor_element?.closest(".katex") &&
+        anchor_element?.closest(".katex-display") === null &&
+        focus_element?.closest(".katex-display") === null
     ) {
         return;
     }
 
-    if (anchor_element) {
-        const parent = anchor_element.closest(".katex");
-        const is_math_block = parent !== null && parent !== selection.anchorNode;
-        if (is_math_block) {
-            const range = document.createRange();
-            range.selectNodeContents(parent);
-            selection.removeAllRanges();
-            selection.addRange(range);
+    const is_single_expression =
+        focus_element?.closest(".katex") === anchor_element?.closest(".katex");
+    if (is_single_expression) {
+        if (anchor_element) {
+            const parent = anchor_element.closest(".katex");
+            const is_expression = parent !== null && parent !== selection.anchorNode;
+            if (is_expression) {
+                const range = document.createRange();
+                range.selectNodeContents(parent);
+                selection.removeAllRanges();
+                selection.addRange(range);
+            }
+        } else if (focus_element) {
+            const parent = focus_element.closest(".katex");
+            const is_expression = parent !== null && parent !== selection.focusNode;
+            if (is_expression) {
+                const range = document.createRange();
+                range.selectNodeContents(parent);
+                selection.removeAllRanges();
+                selection.addRange(range);
+            }
         }
-    } else if (focus_element) {
-        const parent = focus_element.closest(".katex");
-        const is_math_block = parent !== null && parent !== selection.focusNode;
-        if (is_math_block) {
-            const range = document.createRange();
-            range.selectNodeContents(parent);
-            selection.removeAllRanges();
-            selection.addRange(range);
+    } else {
+        const range = selection.getRangeAt(0);
+        assert(anchor_element !== null && focus_element !== null);
```

Comment:
Why is this assert statement true? We returned above if both of these were null, but what if only one is? Is that an impossible situation?